### PR TITLE
[SYCLomatic] Remove volatile from vector method

### PR DESCRIPTION
**Pull request by 0x12CC**
Signed-off-by: Michael Aziz <michael.aziz@intel.com>

### CODE REVIEW

**Review by LU-JOHN — APPROVED** [review]
LGTM

File: clang/lib/DPCT/ASTTraversal.cpp [hunk 1]
```diff
@@ -3598,6 +3598,20 @@ void VectorTypeMemberAccessRule::runRule(
   }
 
   if (auto ME = getNodeAsType<MemberExpr>(Result, "DerivedVecMemberExpr")) {
+    const auto *MD = DpctGlobalInfo::findAncestor<CXXMethodDecl>(ME);
+    if (MD && MD->isVolatile() && ME->isImplicitAccess()) {
```
Comment (ziranzha) [inline clang/lib/DPCT/ASTTraversal.cpp]: Will case like `this->x` be handled?

Comment (0x12CC) [reply clang/lib/DPCT/ASTTraversal.cpp]: I've updated the migration to handle `this->x` and added more test cases.

**Comment by keryell** [comment]
Just curious: why this SYCL implementation does not support `volatile`?

**Comment by 0x12CC** [comment]
> Just curious: why this SYCL implementation does not support `volatile`?

This SYCL specification doesn't include volatile-qualifed declarations for these methods. The implementation does not include them to match the specification.

**Comment by keryell** [comment]
Sure, the `volatile` keyword does not appear in the specification. But it does not prevent to ignore or implement it to do nothing more.
But I am curious also about the use case in some CUDA applications at the first place.

**Comment by jzc** [comment]
If you wanted to access the components of a vector type in a volatile way in SYCL, how about 
```
*const_cast<volatile int*>(&const_cast<Derived*>(this)->x))
```
i.e. after removing volatility to be able to call the accessor, dereference a pointer to the component with volatility added? At least for CPU code, this seems to work (https://godbolt.org/z/axnfx65dd) - when `foo` is given a nonvolatile argument, the read is optimized out, while with a volatile argument, the load is still there.

**Review by jzc — APPROVED** [review]
LGTM though, I think its fine to remove the volatility if we emit a warning.

**Pull request merged**